### ISSUE TITLE
Iphone 3D engine with openGL 2.0 experience

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#3379]
Iphone 3D engine with openGL 2.0 experience<br>
<br>
http://bit.ly/2VCxWUT<br>
<br>
About 0 Minutes<br>
<br>
<div><section>

<p>October 19th, 2009</p>
<p>
<a href="http://fabiensanglard.net/dEngine/index.php">
<img src="http://fabiensanglard.net/dEngine/dEngine.jpg"> </a> Pumped up by the tremendous success of <a href="http://itunes.apple.com/WebObjects/MZStore.woa/wa/viewSoftware?id=317959717&amp;mt=8">Fluid</a> (4,000,000+ downloads) and <a href="http://itunes.apple.com/WebObjects/MZStore.woa/wa/viewSoftware?id=312575632&amp;mt=8">Fluid 2</a> (50,000+ downloads), I started two month ago to write a real game for iPhone: Here is a modest report of my experience in the process.</p>
<p>
This entry focus mostly on the renderer, I'm planning on writing more about the toolchain and the visibility surface determination system later.
</p>
<h3>Screenshots</h3>
<h3>A few videos</h3>
<p>
The following videos show characters from Doom3 that I used to test the engine, the HellKnight is 2200 poly, the rest of the room visible is 1000.
The materials all feature a diffuse map, a normal map and a specular map ( up to 512x512 ). The shadow is generated via a shadowmap ( because render to depth texture is not supported on iPhone (<code>GL_OES_depth_texture</code> <u>Edit 2011:</u> As of iOS 4.1, <code>GL_OES_depth_texture</code> <b>IS</b> supported), depth value are packed in a R8G8B8A8 color texture twice the size of the screen).</p>
<p>
</p>
<table>
<tbody>
<tr>
<td>
<b>iPHone 3GS programmable pipeline, running at 27fps.<br></b>
</td>
<td>
<b>iPhone 2G/3G fixed pipeline, running at 45fps.<br><b>
</b></b>
</td>
</tr>
<tr>
<td>
<video width="480" preload="none" poster="doomRotated-poster.jpg" height="336" data-vscid="7xq7bc91a">

Your browser does not support the video tag.
</video>
</td>
<td>
<video width="480" preload="none" poster="doom3GRotated-poster.jpg" height="336" data-vscid="dslggjzwz">

Your browser does </video>
</td>
</tr>
</tbody>
</table>
<br>

<h3>Choices</h3>
<p>
<br>
My early choices were:</p>
<ul>
<li>It's going to be a 3D shmup because it's the type of games I enjoy the most. Nowaday they are not so trendy but I really appreciated the instant fun and beauty in them. They are also relatively simple to program. <a href="http://www.youtube.com/watch?v=ToBdzV7w5Pc">Ikaruga</a> is the quality (at least graphic) I am aiming at.</li>
<li>Run on ANY iphone/iTouch from early 2G to 3GS. Frame/s never goes below 30.</li>
<li>One unique collection of asset for every devices, the same model is used across all plateforms.</li>
<li>On high end iDevice, support shadows, bumpMapping and specular mapping.</li>
<li>Material centric, allowing high end device to use PNG, others PVRT.</li>
<li>Ability to "scale" automatically depending on the plateform detected.</li>
<li>Only one codebase (fat binary used for Cortex A8 Neon SIMD/ 3GS optimizations).</li>	<li>Written in pure C.</li>
</ul>
<p>
I decided to go with C because:
</p>
<ul>
<li>I like to be close to the metal and it's very valuable when you program on a mobile plateform. On a very limited environment, the least the language "take care for me", the happier I am.</li>
<li>I enjoy this language very much and it's paramount for an indy project to jump on anything that will helps you to remain motivated.</li>
<li>It's easier (at least for me )to write portable code and I wanted to be able to run the engine on Windows in order to allow artist using Vista or XP to join the project anyway (and it does run on XP now).</li>
</ul>
<br>
<h3>C is beautiful</h3>
<p>
Writing an engine in C is slow mainly for two reasons:<br>
</p>
<ul>
<li>You lose access to plenty of goodies such as &lt;string&gt;, &lt;vector&gt; or &lt;map&gt;.</li>
<li>You lose operator overloading which is handy when you work on vectors</li>
<li>Inheritance is not too much of a big deal but not have polymorphism makes things ugly.</li>
</ul>
<p>
To elaborate on the polymorphism aspect: It was really annoying when trying to think of an elegant way to have two renderers: One for the fixed pipepline used on 2G/3G and iTouch, the other for the programmable pipepline on 3GS. An interface or an abstract class would have been perfect but it was not an option. I ended up reusing a pattern that I found while reading the Quake 2 source code: An object containing function pointers, initialized at runtime.</p>
<pre>
	typedef struc renderer_t
	{
		uchar type;
		void (*Set3D)(void); 
		void (*StartRendition  )(void); 
		void (*StopRendition  )(void); 
		void (*SetTexture)(unsigned int);
		void (*RenderEntities)(void);
		void (*UpLoadTextureToGpu)(texture_t* texture);
		void (*Set2D)(void);
		//...
	} renderer_t

				
	//	renderer_fixed.h
	void initFixedRenderer(renderer_t* renderer);

	//   renderer_progr.h
	void initProgrRenderer(renderer_t* renderer);

				</pre>	<p>
The "implementation" of every function is hidden in the .c of each renderer, <code>initFixedRenderer</code> and <code>initProgrRenderer</code> only expose the function address via the pointer.<br>
In the end I think it's pretty neat and besides this I think C is a really good choice to write your engine on iPhone/iTouch, I recommend it.<br></p>
<h3>Degrading....</h3>
<p>	Learning 3D on an NVidia 8600 GT and then moving to a PowerVR SGX/MBX made me feel like a spoiled brat. It hit me reality hard with the first version of the fixed pipeline running at 7 fps for 4000 polygons displayed. My first reaction was of course to blame it on the hardware, wondering how John Carmack could have compared the power of an iPhone to a Dreamcast.
I demoed it to my French exiled bro Vinc and even his legendary positive attitute had no come back to such a failure.</p>
<p>
So I read a lot, studied, tried, trashed and I made my way into iPhone and powerVR philosophy:</p>
<p> It turns out you can do A LOT with this hardware but you have to take into account some of it's limitations:</p>
<p>
My first mistake was to forget about bandwidth consumption. Modern GPU have dedicated VRAM with double digits Gb/s output. The iPhone is far from this and the first thing to do is to tune down the size (and quality) of what you send to the GPU:</p>
<p>
<u>Texture compression</u> is a big win as a 32bits per texel RGBA textures is a pig with no real reason to exist when working with a small display. OpenGS ES 1.1 and 2.0 do not require the GPU to support any texture compression but the good guys at Imagination Technologies provided support for PVRTC which bring down consumption to as low at 2bits per pixel with alpha support !</p>
<p>
<u>Vertex metadatas</u> can be slimmed down as well:</p>
<p>
A "regular" vertex is:</p>
<br>
<pre>
	Vertex Elementary unit:

	position   	3 floats
	normal		3 floats
	tangent		3 floats
	textureCoo	2 floats
	-------------------
	            44 bytes

</pre>
<p>By packing the components in "shorts" instead of "floats" via normalization, you end up having:<br></p>
<pre>
	Vertex Elementary unit:

	position   	3 floats
	normal		3 shorts
	tangent		3 shorts
	textureCoo	2 shorts
	-------------------
	            28 bytes

</pre>
<p>It's almost like we "compress" the data on the CPU, send it to the GPU where they are "decompressed". Abusing normalization divide bandwidth consumption by almost 50% and help to slightly improve performances.</p>
<p>
<u>Compiler tuning</u> is also important. Xcode is setup by default to generate ARM binaries using the Thumb instruction set, which is 16 bits instead of 32 bits. This reduce the size of the binary and the cost for Apple but it's bad for 3D as Thumb instruction have to be translated to 32bits.
Uncheck this option for an instant gain of performances.</p>
<p>
<u>Framebuffer refresh</u> can also be improved a lot with 3.1 firmware. This is an issue I mentionned in my <a href="http://fabiensanglard.net/wolf3d/index.php">article</a> about Wolfenstein for iPhone: NSTimer is an abomination and I was trilled to find we can now use <code>CADisplayLink</code> to perform vsync and get adaptative framerate ( although I'm experimenting some nasty touchesMoved on non 2G v3.X devices, if you have any info about this, email me !).</p>
<p>
<u>Reduze Framebuffer colorspace</u> is an other way to improve performances by reducing the amount of written data. Move from 24bits color space to 16 bits provides some good improvements.</p>
<pre>    
    CAEAGLLayer *eaglLayer = (CAEAGLLayer *)self.layer;
		
    eaglLayer.opaque = YES;
    eaglLayer.drawableProperties = [NSDictionary dictionaryWithObjectsAndKeys:
        [NSNumber numberWithBool:YES], 
        kEAGLDrawablePropertyRetainedBacking, 
        
        //FTW
        //kEAGLColorFormatRGBA8, 
        kEAGLColorFormatRGB565,

        kEAGLDrawablePropertyColorFormat, nil];
        
	</pre>
<p>
Stating the obvious here, but <u>reduce texture &amp; blending mode</u> switches are very important ( Forget about good perf if you do more than 60 textures changes). The material approach of the engine can very handy in this regard.</p>
<p>
Reduce blending of your polygons is PARAMOUNT: PowerVR performs TBDR (tile-based deferred rendering) which mean that one pixel is rendered only once via hidden surface removal, blending is defeating the purpose. My take is that a blended polygon is rendere regardless of the culling outcome and it destroys perfs.</p>
<p>
And last but not least, <u>optimize the vertice indices</u> so GPU fetches will hit the cache as much as possible.</p>
<br>

<h3>OpenGL ES 2.0 = One GPU + One giant notepad.</h3>
<p>
One other difficulty working with mobile shaders was to discover that OpenGL ES 2.0 is NOT a superset of openGL ES 1.1. This can sound like a minor issue but what it means is that the drivers stop providing you with a bunch of things such as GL_PROJECTION matrix, GL_MODEL_VIEW matrix, GL_TEXTURE matrix and so on. You also lose every verterxArray functions: glNormalPointer,glTexCoordPointer,glVertexPointer. In the shader, contraty to OpenGL 2.0 again nothing is provided (no gl_ModelViewMatrix, no gl_NormalMatrix, no ftransform and no gl_LightSource either. Anything you want to have in the shader must be passed from the CPU via an uniform, a variable or an attribute (vertex position, normal and textCoo INCLUDED !!).</p>
<p>In a nutshell: 2.0 does NOTHING for you, and when I say nothing I mean nada, zip, rien, queues d'ales mon gros.</p>
<p>
This was annoying at first because instead of improving the quality of the rendition right when you jump on 2.0, you actually take a few steps back. This disapointment was fast replaced by the satisfaction to have total access to the GPU. This unlock the ability to perform all bumpMapping illumination in model space instead of the traditional eye space (slightly faster).</p>
<br>
The other thing was the satisfaction to have programmed something totally empty when it was provided. This is super nerdy but there is nothing like feeling that you really "get" the hardware.
<p>
I've also tried to experiment with an UberShader design: Besides the shadowmap generator there is only one shader in the engine, effects (shadow, spec mapping, bumpmapping) are activated via openGL pre-processing during compile time:</p>
<pre>
    //snipet of the fragment shader

    #ifdef BUMP_MAPPING
        bump		=  texture2D(s_bumpMap, v_texcoord).rgb * 2.0 - 1.0;
        lamberFactor  =  max(0.0,dot(lightVec, bump) );
        specularFactor = max(0.0,pow(dot(halfVec,bump),materialShininess)) ;
    #else
        lamberFactor  =  max(0.0,dot(lightVec, v_normal) );
        specularFactor = max(0.0,pow(dot(halfVec,v_normal),materialShininess)) ;
    #endif
	
    #ifdef SPEC_MAPPING
        vec3 matTextColor = texture2D(s_specularMap, v_texcoord).rgb; 
    #else
        vec3 matTextColor = matColorSpecular;
    #endif

</pre>
<p>
Depending on the materials properties used in a scene , the shader is re-compiled at runtime and then cached. This approach allow to reduce branching operation in the shader. I was very pleased with the result, if I stay below 10/15 shader switches per frame there is no significant performance drop.<br>
</p> <p>
</p>
<h3>Try, Trash, Repeat....</h3>
<p>
The most frustrating part is to reach a point where components and the overall design start to feel "not right". It generally happen after a serie of "patchwork" when you get to implement things you hadn't anticipated from the beginning.<br>
It's generally a good time to "start over", keeping your math and loader library (so you actually don't redo the boring parts). That may sound like a massive overshot but I found it quite efficient.</p>
<p>
Within two months I had to trash numerous components, pretty much everything was written two sometimes three times. This is of course not realizable if you work in a team or with a lot of legacy but starting from scratch was a luxury I was glad to afford myself.</p>
<br>

<h3>3GS Neon and SIMD </h3>
<p>
Trying to push the 3GS Cortex-A8 to the maximum with SIMD instruction provided no gain. The CPU does perform vectorial operations but not enough to show improvements while using SIMD.<br>
I found the ability to generate a fat binary armv6/armv7 via Xcode's (Optimized armv6 armv7) option quite neat: Just wrap the maths functions with <code>#ifdef __ARM_NEON__</code> and you will have no overhead at runtime.</p>
<p> Thanks to Justin Sauders for sharing his <a href="http://code.google.com/p/oolongengine/source/browse/trunk/Oolong+Engine2/Math/neonmath/neon_matrix_impl.cpp?spec=svn143&amp;r=143">NEON/ARM/SIMD</a> library. </p>	<h3>The shadow problem</h3>
<p>
<u>Edit 2011:</u> As of iOS 4.1, <code>GL_OES_depth_texture</code> <b>IS</b> supported, you can render directly to a depth buffer.</p>
<p>
I love shadows effects, I think the realism and ambiance you get totally justify the cycles and bandwidth cost. It doesn't come for free in openGL and it's quite ugly to do with the fixed pipeline but I was trilled to have it working on mobile shaders. Unfortunatly as of today, iPhones don't support <code>GL_OES_depth_texture</code>, which mean you cannot render directly to the a depth texture. The workaround is to pack a 32 floating point value into 4x4 bytes color (RGBA) texture:</p>
<br>
<pre>
	// This is the shadowmap generator shader

	const vec4 packFactors = vec4( 256.0 * 256.0 * 256.0,256.0 * 256.0,256.0,1.0);
	const vec4 bitMask     = vec4(0.0,1.0/256.0,1.0/256.0,1.0/256.0);

	void main(void)
	{
		float normalizedDistance  = position.z / position.w;
		normalizedDistance = (normalizedDistance + 1.0) / 2.0;

		vec4 packedValue = vec4(fract(packFactors*normalizedDistance));
		packedValue -= packedValue.xxyz * bitMask;

		gl_FragColor  = packedValue;
	}

			</pre>
<p>
This method to pack float in bytes is pretty clever (not mine) because it accounts for the internal accuracy of any GPU ( via the substraction line) and hence can be used on any kind of GPU (PowerVR,ATI,NVidia). Gratz to however came up with this.</p>
<h3>And now what ? </h3>
<p>
A few technical things ( see what can be done to improve the shadows with Variance &amp;| Percentage Closer filtering) and improve the fixed pipeline with DOT3 bumpMapping or shadow but mainly:</p>
<p>
Tools</p>
<p>
To write a good engine is a first step to implement something of quality but if artists cannot easily input their creation it's like you have done nothing.</p>
<br>
Doom3 is a good example of a fantastic piece of technology that was impaired during licensing by the need for text files and commandline utility ( besides the difficuly to render outdoor environments ).	Tools are not necessarily rocket science but they take a lot of time because they need to have a good friendy GUI.<p>
I am going to keep working my ass off because shipping a game has been a dream for a long time now. Hopefully a talented and motivated artist will cross the road with this piece of code.</p>
<p>
I hope I will finish it.</p>
<h2>Comments</h2>

<p>Fabien Sanglard @2009</p>
</section></div><br>
<br>
<br>
<br>
Ninja<br>
<br>
via fabiensanglard.net http://bit.ly/rNP4ZU<br>
<br>
April 11, 2019 at 02:29AM
